### ISSUE TITLE
extending double-clicked selection with shift+RMB deselects part of original selection

### ISSUE THREAD

**Issue by p1otr** [alacritty/alacritty#4237]
Hi,

I'm testing Alacritty as an alternative terminal for URxvt (which is not Wayland ready yet) and I like it so far. Thanks!

There's one¹ annoyance, though: when I double or triple click on some text and later want to extend selection with a right mouse button click - original selection starts on the character I double/triple clicked instead of the whole word.

F.e. if I double click on "a" from "bar" in "foo bar baz" - "bar" is selected as expected.
When I right-click on "f" - "foo ba" is selected instead of expected "foo bar".
Right-clicking later on "z" selects "r baz" instead of extending previous extension (i.e. I expect to have "foo bar baz" selected after second click).

Please let me know if it's configurable and just didn't find the right switch.
Thanks again for writing Alacritty!

[¹] well… two, but from what I know there's nothing you can do about bell (`echo -e '\a'`) not setting urgent flag for windows on Wayland - as there no such thing there yet

### System

OS: Linux (Debian Sid)
Version: 0.6.0-dev (1bf40a8)
Linux/BSD: Wayland, Sway

**Comment by chrisduerr** [comment]
If you want to semantically extend the selection, you need to double right click. If you extend it with just a single click, it will be converted to a normal selection.

**Comment by p1otr** [comment]
that's better than what I was doing before, but when I extend selection - I usually want to fine tune it so another double click is not always optimal (it selects more than I want). Thanks for the tip, though

**Comment by kchibisov** [comment]
I see. I'm not sure if that's behavior we want to have, since the goal is to change selections on a fly wrt original selection. So the span from triple click selection will include the entire line, so single right click selection will include that line + some minor part of it. What do you think @chrisduerr ?

**Comment by chrisduerr** [comment]
> What do you think @chrisduerr ?

I think that if you don't want a semantic selection, you probably shouldn't start with one?

**Comment by kchibisov** [comment]
>I think that if you don't want a semantic selection, you probably shouldn't start with one?

I guess, but I feel like the way our expansion works isn't very intuitive, since you can change originally selected things, which isn't desired and may require users some time to understand how it works? So the way expanding part works is fine, however the selected region you've done at first should be stored in expanded mode? On the other side the way it's done that way right now is due to Vi mode, where those changes are following vim.

**Comment by freshprince** [comment]
One of my usual selection movements with the mouse are double-click inside a word to make the selection borders semantic and then right click to expand the selection to somewhere that might not be a semantic border.  It feels very weird when this action besides of expanding the selection also shrinks it at the other end.

I can see that this behaviour seems reasonable when I look at the code but maybe it would make sense to create a transition function when switching selection types that introduces the expected behaviour that also honours Vi mode. If that would be something you might consider I could try my luck. What do you think?

**Comment by chrisduerr** [comment]
I don't think this would be a good issue for contributors since it requires some major changes.

**Comment by nixpulvis** [comment]
I believe the solution to this would remove the concept of normal, semantic and line selection from the selection itself, and put it on some part of the process of selecting. Once I have made a selection it should not matter *how* I made that selection.

**Comment by markstos** [comment]
I ran into a related issue today with 0.8.0, I'm not sure if it's the same. In Vi mode, with a Normal Selection in progress, I tried key bindings for "First" and "Last", expecting the selection be extended. The cursor did move to the first and last positions, but the selection was dropped. 

My bindings were:

```
- { key: Space,                     mode: Vi, action: ToggleNormalSelection   }
- { key: Home,                      mode: Vi, action: First                   }
- { key: End,                         mode: Vi, action: Last                    }
````

**Comment by anarcat** [comment]
> If you want to semantically extend the selection, you need to double right click. If you extend it with just a single click, it will be converted to a normal selection.

that is a fascinating approach that breaks with decades of left/right-click primary selection semantics going back to X11 standards... i thought it was just not possible to extend word selection before finding this issue and that comments.

i would suggest implementing some sort of "mode" that would "keep" the word selection active even if one doesn't "double-right-click. 

i don't know about you, but my "double-clicking" skills have always been bound to left click, i don't remember ever having to "double-right-click" in my life before today. :)
